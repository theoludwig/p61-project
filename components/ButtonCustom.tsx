import type { StyleProp, ViewStyle } from "react-native"
import { Pressable, StyleSheet, Text } from "react-native"

export interface ButtonCustomProps
  extends React.ComponentProps<typeof Pressable> {
  children: React.ReactNode
  style?: StyleProp<ViewStyle>
}

export const ButtonCustom: React.FC<ButtonCustomProps> = (props) => {
  const { children, style, ...rest } = props

  return (
    <Pressable style={[styles.button, style]} {...rest}>
      <Text style={styles.text}>{children}</Text>
    </Pressable>
  )
}

const styles = StyleSheet.create({
  button: {
    padding: 10,
    borderWidth: 1,
    borderColor: "black",
    borderRadius: 10,
    marginTop: 10,
    backgroundColor: "#152B5D",
  },
  text: {
    color: "white",
    textAlign: "center",
    fontSize: 15,
  },
})
