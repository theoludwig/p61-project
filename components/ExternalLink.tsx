import { Link } from "expo-router"
import * as WebBrowser from "expo-web-browser"
import { Platform } from "react-native"

type LinkProps = React.ComponentProps<typeof Link>

export const ExternalLink: React.FC<
  Omit<LinkProps, "href"> & {
    href: string
  }
> = (props) => {
  const { href, ...rest } = props

  return (
    <Link
      target="_blank"
      href={href as unknown as LinkProps["href"]}
      {...rest}
      onPress={async (event) => {
        if (Platform.OS !== "web") {
          event.preventDefault()
          await WebBrowser.openBrowserAsync(href)
        }
      }}
    />
  )
}
