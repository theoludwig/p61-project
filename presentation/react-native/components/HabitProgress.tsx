import { useState } from "react"
import { ScrollView, StyleSheet, View } from "react-native"
import { Button, Snackbar, Text, TextInput } from "react-native-paper"
import { SafeAreaView } from "react-native-safe-area-context"

import type { GoalNumeric } from "@/domain/entities/Goal"
import { GoalNumericProgress } from "@/domain/entities/Goal"
import type { HabitHistory } from "@/domain/entities/HabitHistory"
import { useHabitsTracker } from "@/presentation/react/contexts/HabitsTracker"
import { LOCALE, capitalize } from "@/utils/strings"

export interface HabitProgressProps {
  habitHistory: HabitHistory
  selectedDate: Date
}

export const HabitProgress: React.FC<HabitProgressProps> = ({
  habitHistory,
  selectedDate,
}) => {
  const { habitsTrackerPresenter, habitGoalProgressUpdate } = useHabitsTracker()

  const [isVisibleSnackbar, setIsVisibleSnackbar] = useState(false)

  const onDismissSnackbar = (): void => {
    setIsVisibleSnackbar(false)
  }

  const goalProgress = habitHistory.getGoalProgressByDate(selectedDate)

  const values = {
    progress: 0,
    min: 0,
    max: 0,
  }
  if (goalProgress.isNumeric()) {
    values.max = goalProgress.goal.target.value
  }
  const [progressValue, setProgressValue] = useState(values.progress)

  if (!goalProgress.isNumeric()) {
    return <></>
  }

  const progressTotal = goalProgress.progress + progressValue

  const handleSave = async (): Promise<void> => {
    setIsVisibleSnackbar(true)
    await habitsTrackerPresenter.habitUpdateProgress({
      date: selectedDate,
      habitHistory,
      goalProgress: new GoalNumericProgress({
        goal: habitHistory.habit.goal as GoalNumeric,
        progress: progressValue,
      }),
    })
    setProgressValue(0)
  }

  return (
    <SafeAreaView style={{ flex: 1, justifyContent: "space-between" }}>
      <ScrollView
        contentContainerStyle={{
          justifyContent: "center",
          alignItems: "center",
          paddingHorizontal: 20,
        }}
      >
        <Text
          style={{
            fontWeight: "bold",
            fontSize: 28,
            textAlign: "center",
          }}
        >
          {habitHistory.habit.name}
        </Text>

        <Text
          style={{
            marginTop: 10,
            fontWeight: "bold",
            fontSize: 18,
            textAlign: "center",
          }}
        >
          {capitalize(habitHistory.habit.goal.frequency)} Progress
        </Text>

        <Text
          style={{
            fontSize: 16,
            textAlign: "center",
            marginBottom: 15,
          }}
        >
          {selectedDate.toLocaleDateString(LOCALE, {
            weekday: "long",
            year: "numeric",
            month: "long",
            day: "numeric",
          })}
        </Text>

        <View
          style={{
            width: "100%",
            borderBottomWidth: 1,
            borderColor: "#f57c00",
            marginVertical: 10,
          }}
        />

        <Text style={{ marginVertical: 10, fontWeight: "bold", fontSize: 18 }}>
          {goalProgress.progress.toLocaleString()} /{" "}
          {goalProgress.goal.target.value.toLocaleString()}{" "}
          {goalProgress.goal.target.unit}
        </Text>

        <TextInput
          placeholder="Progress to add (e.g: 5 000)"
          value={progressValue === 0 ? "" : progressValue.toString()}
          onChangeText={(text) => {
            const hasDigits = /\d+$/.test(text)
            if (text.length <= 0 || !hasDigits) {
              setProgressValue(0)
              return
            }
            setProgressValue(Number.parseInt(text, 10))
          }}
          style={[
            styles.spacing,
            {
              width: "80%",
            },
          ]}
          mode="outlined"
          keyboardType="numeric"
        />

        {goalProgress.progress > 0 && progressValue > 0 ? (
          <Text
            style={{
              fontSize: 16,
              textAlign: "center",
              marginBottom: 15,
            }}
          >
            {goalProgress.progress.toLocaleString()} +{" "}
            {progressValue.toLocaleString()} = {progressTotal.toLocaleString()}{" "}
            {goalProgress.goal.target.unit}
          </Text>
        ) : (
          <></>
        )}

        <Button
          mode="contained"
          onPress={handleSave}
          loading={habitGoalProgressUpdate.state === "loading"}
          disabled={
            habitGoalProgressUpdate.state === "loading" || progressValue === 0
          }
          style={[styles.spacing, { width: "80%" }]}
        >
          Save Progress ✨
        </Button>

        <View
          style={{
            width: "100%",
            borderBottomWidth: 1,
            borderColor: "#f57c00",
            marginVertical: 10,
          }}
        />
      </ScrollView>

      <Snackbar
        visible={isVisibleSnackbar}
        onDismiss={onDismissSnackbar}
        duration={2_000}
      >
        ✅ Habit Saved successfully!
      </Snackbar>
    </SafeAreaView>
  )
}

const styles = StyleSheet.create({
  spacing: {
    marginVertical: 16,
  },
})
