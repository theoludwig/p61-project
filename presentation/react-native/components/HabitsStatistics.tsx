import { Card, Divider, Text } from "react-native-paper"
import CircularProgress from "react-native-circular-progress-indicator"
import { Agenda } from "react-native-calendars"
import { useState } from "react"

import { getNowDateUTC, getISODate } from "@/utils/dates"
import type { HabitsTracker } from "@/domain/entities/HabitsTracker"
import { LOCALE } from "@/utils/strings"
import {
  GOAL_FREQUENCIES,
  GOAL_FREQUENCIES_TYPES,
} from "@/domain/entities/Goal"
import { calculateRatio } from "@/utils/maths"

export interface HabitsStatisticsProps {
  habitsTracker: HabitsTracker
}

export const HabitsStatistics: React.FC<HabitsStatisticsProps> = (props) => {
  const { habitsTracker } = props

  const today = getNowDateUTC()
  const todayISO = getISODate(today)

  const [selectedDate, setSelectedDate] = useState<Date>(today)
  const selectedDateISO = getISODate(selectedDate)

  return (
    <Agenda
      firstDay={1}
      showClosingKnob
      onDayPress={(date) => {
        setSelectedDate(new Date(date.dateString))
      }}
      markedDates={{
        [todayISO]: { marked: true, today: true },
      }}
      maxDate={todayISO}
      selected={selectedDateISO}
      renderList={() => {
        return (
          <>
            <Divider />

            <Text
              style={{
                fontWeight: "bold",
                fontSize: 22,
                textAlign: "center",
                marginVertical: 10,
              }}
            >
              {selectedDate.toLocaleDateString(LOCALE, {
                weekday: "long",
                year: "numeric",
                month: "long",
                day: "numeric",
              })}
            </Text>

            {GOAL_FREQUENCIES.map((frequency) => {
              const { totalGoalsSuccess, totalGoals } =
                habitsTracker.getHabitsStatisticsByDateAndFrequency({
                  selectedDate,
                  frequency,
                })
              const percentage =
                calculateRatio(totalGoalsSuccess, totalGoals) * 100
              return {
                totalGoalsSuccess,
                totalGoals,
                percentage,
                frequency,
              }
            })
              .filter(({ totalGoals }) => {
                return totalGoals > 0
              })
              .map(
                ({ frequency, totalGoals, totalGoalsSuccess, percentage }) => {
                  return (
                    <Card
                      key={frequency}
                      mode="elevated"
                      style={{ marginVertical: 8, marginHorizontal: 10 }}
                    >
                      <Card.Content
                        style={{
                          justifyContent: "center",
                          alignItems: "center",
                        }}
                      >
                        <Text variant="bodyMedium" style={{ marginBottom: 5 }}>
                          {totalGoalsSuccess} achieved goals in the{" "}
                          {GOAL_FREQUENCIES_TYPES[frequency]} out of{" "}
                          {totalGoals}.
                        </Text>
                        <CircularProgress
                          value={percentage}
                          progressValueColor={"#ecf0f1"}
                          circleBackgroundColor="black"
                          titleColor="white"
                          title="%"
                        />
                      </Card.Content>
                    </Card>
                  )
                },
              )}
          </>
        )
      }}
    />
  )
}
