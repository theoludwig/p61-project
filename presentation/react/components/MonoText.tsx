import { StyleSheet, Text } from "react-native"

export const MonoText: React.FC<Text["props"]> = (props) => {
  const { style, ...rest } = props

  return <Text style={[style, styles.text]} {...rest} />
}

const styles = StyleSheet.create({
  text: {
    fontFamily: "SpaceMono",
  },
})
