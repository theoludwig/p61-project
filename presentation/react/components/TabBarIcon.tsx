import FontAwesome from "@expo/vector-icons/FontAwesome"

/**
 * @see https://icons.expo.fyi/
 * @param props
 * @returns
 */
export const TabBarIcon: React.FC<{
  name: React.ComponentProps<typeof FontAwesome>["name"]
  color: string
}> = (props) => {
  return <FontAwesome size={28} style={{ marginBottom: -3 }} {...props} />
}
