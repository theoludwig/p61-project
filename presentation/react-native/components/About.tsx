import { View } from "react-native"
import { Text } from "react-native-paper"
import { SafeAreaView } from "react-native-safe-area-context"

import { ExternalLink } from "@/presentation/react-native/ui/ExternalLink"
import { getVersion } from "@/utils/version"

export interface AboutProps {
  actionButton: React.ReactNode
  footer?: React.ReactNode
}

export const About: React.FC<AboutProps> = (props) => {
  const { actionButton, footer } = props

  const version = getVersion()

  return (
    <SafeAreaView
      style={{
        flex: 1,
        paddingHorizontal: 20,
        justifyContent: "center",
      }}
    >
      <View
        style={{
          alignItems: "center",
          marginVertical: 20,
        }}
      >
        <Text
          style={{
            fontWeight: "bold",
            fontSize: 28,
            textAlign: "center",
          }}
        >
          Habits Tracker
        </Text>

        <Text
          style={{
            marginTop: 6,
            fontWeight: "bold",
            fontSize: 18,
            textAlign: "center",
          }}
        >
          To perform at work and in everyday life.
        </Text>

        <Text
          style={{
            marginTop: 6,
            fontWeight: "bold",
            fontSize: 16,
            textAlign: "center",
          }}
        >
          v{version}
        </Text>
      </View>

      <Text variant="bodyLarge" style={{ textAlign: "center" }}>
        <ExternalLink href="https://unistra.fr" style={{ color: "#006CFF" }}>
          Universit√© de Strasbourg
        </ExternalLink>
      </Text>
      <Text variant="bodyLarge" style={{ textAlign: "center" }}>
        BUT Informatique - IUT Robert Schuman
      </Text>
      <Text variant="bodyLarge" style={{ textAlign: "center" }}>
        P61 Mobile Development
      </Text>

      {footer}

      <View
        style={{
          justifyContent: "center",
          alignItems: "center",
          marginVertical: 20,
        }}
      >
        {actionButton}
      </View>
    </SafeAreaView>
  )
}
