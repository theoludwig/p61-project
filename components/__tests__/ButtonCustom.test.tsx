import renderer from "react-test-renderer"

import { ButtonCustom } from "@/components/ButtonCustom"

describe("<ButtonCustom />", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(<ButtonCustom>Awesome Button!</ButtonCustom>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
