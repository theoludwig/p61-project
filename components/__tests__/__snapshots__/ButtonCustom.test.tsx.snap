// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonCustom /> renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "backgroundColor": "#152B5D",
        "borderColor": "black",
        "borderRadius": 10,
        "borderWidth": 1,
        "marginTop": 10,
        "padding": 10,
      },
      undefined,
    ]
  }
>
  <Text
    style={
      {
        "color": "white",
        "fontSize": 15,
        "textAlign": "center",
      }
    }
  >
    Awesome Button!
  </Text>
</View>
`;
